staging/src/k8s.io/apiserver/plugin/pkg/kmsprovider/vault/client_test.go

```diff
@@ -0,0 +1,419 @@
+}
+
+func customAuthPathServer(t *testing.T, auth string) *httptest.Server {
+	handlers := DefaultTestHandlers(t)
+
+	// Replace with custom auth path
+	for _, key := range []string{"/v1/auth/cert/login", "/v1/auth/approle/login"} {
+		newKey := strings.Replace(key, "auth", auth, 1)
+		handlers[newKey] = handlers[key]
+		delete(handlers, key)
+	}
+
+	return VaultTestServer(t, handlers)
+}
+
+func TestForbiddenRequest(t *testing.T) {
+	forbiddenFunc := func(w http.ResponseWriter, r *http.Request) {
+		http.Error(w, http.StatusText(http.StatusForbidden), http.StatusForbidden)
+	}
+
+	handlers := map[string]http.Handler{
+		"/v1/transit/encrypt/": http.HandlerFunc(forbiddenFunc),
+		"/v1/transit/decrypt/": http.HandlerFunc(forbiddenFunc),
+	}
+	server := VaultTestServer(t, handlers)
+	defer server.Close()
+
+	config := &EnvelopeConfig{
+		Token:   uuid.NewRandom().String(),
+		Address: server.URL,
+		CACert:  cafile,
+	}
+	client, err := newClientWrapper(config)
+	if err != nil {
+		t.Fatalf("fail to initialize Vault client: %s", err)
+	}
+
+	_, err = client.encrypt("key", "plain")
+	if err == nil {
+		t.Error("error should not be nil for 403 response")
+	}
+	forbidden, ok := err.(*forbiddenError)
+	if !ok {
+		t.Error("forbidden error should be return for 403 response")
+	}
+	if forbidden.err == nil {
+		t.Errorf("error forbidden value, %+v", forbidden)
+	}
+
+	_, err = client.decrypt("key", "cipher")
+	if err == nil {
+		t.Error("error should not be nil for 403 response")
+	}
+	forbidden, ok = err.(*forbiddenError)
+	if !ok {
+		t.Error("forbidden error should be return for 403 response")
+	}
+	if forbidden.err == nil {
+		t.Errorf("error forbidden value, %+v", forbidden)
+	}
+}
+
+func TestRefreshToken(t *testing.T) {
+	server := VaultTestServer(t, nil)
+	defer server.Close()
+
+	configTLS := &EnvelopeConfig{
+		ClientCert: clientCert,
+		ClientKey:  clientKey,
+		Address:    server.URL,
+		CACert:     cafile,
+	}
+	configRole := &EnvelopeConfig{
+		RoleID:  uuid.NewRandom().String(),
+		Address: server.URL,
+		CACert:  cafile,
+	}
+
+	configs := []struct {
+		name   string
+		config *EnvelopeConfig
+	}{
+		{"AuthByTLS", configTLS},
+		{"AuthByAppRole", configRole},
+	}
+
+	for _, testCase := range configs {
+		client, err := newClientWrapper(testCase.config)
+		if err != nil {
+			t.Fatalf("name: %s, fail to initialize Vault client: %s", testCase.name, err)
+		}
+
+		// Refresh token successfully.
+		err = client.refreshToken(testCase.config)
+		if err != nil {
+			t.Fatalf("name: %s, unexpecte error when refresh token: %s", testCase.name, err)
+		}
+	}
+}
+
+func VaultTestServer(tb testing.TB, handlers map[string]http.Handler) *httptest.Server {
+	mux := http.NewServeMux()
+	if handlers == nil {
+		handlers = DefaultTestHandlers(tb)
+	}
+	for path, handler := range handlers {
+		mux.Handle(path, handler)
+	}
+	server := httptest.NewUnstartedServer(mux)
+
+	cert, err := tls.LoadX509KeyPair(serverCert, serverKey)
```

Comment:
We've got some utils you might be able to use here https://godoc.org/k8s.io/client-go/util/cert